pkg/messaging/direct_messaging.go

```diff
@@ -145,6 +157,46 @@ func (d *directMessaging) invokeWithRetry(
 	fn func(ctx context.Context, appID, namespace, appAddress string, req *invokev1.InvokeMethodRequest) (*invokev1.InvokeMethodResponse, error),
 	req *invokev1.InvokeMethodRequest,
 ) (*invokev1.InvokeMethodResponse, error) {
+	// TODO: Once resiliency is out of preview, we can have this be the only path.
+	if d.isResiliencyEnabled {
+		if !d.resiliency.PolicyDefined(app.id, resiliency.Endpoint) {
+			retriesExhaustedPath := false // Used to track final error state.
+			nullifyResponsePath := false  // Used to track final response state.
+			policy := d.resiliency.EndpointPolicy(ctx, app.id, app.address, true)
+			var resp *invokev1.InvokeMethodResponse
+			err := policy(func(ctx context.Context) (rErr error) {
+				retriesExhaustedPath = false
+				resp, rErr = fn(ctx, app.id, app.namespace, app.address, req)
+				if rErr == nil {
+					return nil
+				}
+
+				code := status.Code(rErr)
+				if code == codes.Unavailable || code == codes.Unauthenticated {
+					_, connerr := d.connectionCreatorFn(context.TODO(), app.address, app.id, app.namespace, false, true, false)
+					if connerr != nil {
+						nullifyResponsePath = true
+						return backoff.Permanent(connerr)
+					}
+					retriesExhaustedPath = true
+					return rErr
+				}
+				return backoff.Permanent(rErr)
```

Comment:
Is it right to `return backoff.Permanent()` even when `retriesExhaustedPath` is `false`?

Reply:
Yes, the reason we do that is because we actually don't retry everything. Here's the original retry method:

```go
for i := 0; i < numRetries; i++ {
	resp, err := fn(ctx, app.id, app.namespace, app.address, req)
	if err == nil {
		return resp, nil
	}
	log.Debugf("retry count: %d, grpc call failed, ns: %s, addr: %s, appid: %s, err: %s",
		i+1, app.namespace, app.address, app.id, err.Error())
	time.Sleep(backoffInterval)

	code := status.Code(err)
	if code == codes.Unavailable || code == codes.Unauthenticated {
		_, connerr := d.connectionCreatorFn(context.TODO(), app.address, app.id, app.namespace, false, true, false)
		if connerr != nil {
			return nil, connerr
		}
		continue
	}
	return resp, err
}
```

You can see we only call `continue` when we've gotten `Unavailable` or `Unauthenticated` and were able to reconnect. If it wasn't one of those two errors, we exit from the loop immediately. That's what the permanent error does.